test/sharness/t0054-dag-car-import-export.sh

```diff
@@ -4,22 +4,169 @@
 test_description="Test car file import/export functionality"
 
 . lib/test-lib.sh
+export -f ipfsi
 
+set -o pipefail
+
+reset_blockstore() {
+  node=$1
+  ipfsi $1 pin ls --quiet --type=recursive | ipfsi $1 pin rm &>/dev/null
+  ipfsi $1 repo gc &>/dev/null
+
+  test_expect_success "pinlist empty" '
+    test -z "$( ipfsi $1 pin ls )"
+  '
+  test_expect_success "nothing left to gc" '
+    test -z "$( ipfsi $1 repo gc )"
+  '
+}
+
+# hammer with concurrent gc to ensure nothing clashes
+do_import() {
+  node=$1; shift
+
+  bash -c "while [[ -e spin.gc ]]; do ipfsi $node repo gc >>gc_out 2>&1; done" & gc1_pid=$!
+  bash -c "while [[ -e spin.gc ]]; do ipfsi $node repo gc >>gc_out 2>&1; done" & gc2_pid=$!
+
+  ipfsi $node dag import "$@"
+
+  rm spin.gc
+  wait $gc1_pid
+  wait $gc2_pid
+}
+
+run_online_imp_exp_tests() {
+
+  reset_blockstore 0
+  reset_blockstore 1
+
+  echo -e "Pinned root\tbafkqaaa\tsuccess (root specified in .car header without available data)" > basic_import_expected
+  echo -e "Pinned root\tbafy2bzaceaxm23epjsmh75yvzcecsrbavlmkcxnva66bkdebdcnyw3bjrc74u\tsuccess" >> basic_import_expected
+  echo -e "Pinned root\tbafy2bzaced4ueelaegfs5fqu4tzsh6ywbbpfk3cxppupmxfdhbpbhzawfw5oy\tsuccess" >> basic_import_expected
+
+  touch spin.gc
+  test_expect_success "basic import" '
+    do_import 0 \
+      ../t0054-dag-car-import-export-data/combined_naked_roots_genesis_and_128.car \
+      ../t0054-dag-car-import-export-data/lotus_testnet_export_128_shuffled_nulroot.car \
+      ../t0054-dag-car-import-export-data/lotus_devnet_genesis_shuffled_nulroot.car \
+    | sort > basic_import_actual
+  '
+
+  # FIXME - the fact we reliably fail this is indicative of some sort of race...
+  test_expect_failure "concurrent GC did not manage to find anything" '
+    ! [[ -s gc_out ]]
+  '
+  test_expect_success "basic import output as expected" '
+    test_cmp basic_import_expected basic_import_actual
+  '
+
+  reset_blockstore 0
+  reset_blockstore 1
+
+  mkfifo pipe_testnet
+  mkfifo pipe_devnet
+
+  # test that ipfs correctly opens both pipes and deleting them doesn't interfere with cleanup
+  bash -c '
+    sleep 1
+    cat ../t0054-dag-car-import-export-data/lotus_testnet_export_128_shuffled_nulroot.car > pipe_testnet & cat1_pid=$!
+    cat ../t0054-dag-car-import-export-data/lotus_devnet_genesis_shuffled_nulroot.car > pipe_devnet & cat2_pid=$!
+
+    rm pipe_testnet pipe_devnet
+
+    # extra safety valve to kill the cat processes in case something goes wrong
+    bash -c "sleep 60; kill $cat1_pid $cat2_pid 2>/dev/null" &
+  ' &
+
+  touch spin.gc
+  test_expect_success "fifo import" '
+    do_import 0 \
+      pipe_testnet \
+      pipe_devnet \
+      ../t0054-dag-car-import-export-data/combined_naked_roots_genesis_and_128.car \
+    | sort > basic_fifo_import_actual
+  '
+  # FIXME - the fact we reliably fail this is indicative of some sort of race...
```

Comment:
Let's make sure to update this comment. This isn't due to a race, it's because we're creating blocks and aren't sure where they're coming from.

Actually, looking at this, I think I know where they're coming from. We're flushing multiple times, one for each root. If we have multiple roots, that would do it.

Reply:
Done with reference of issue. I will cross-post the test to be fixed in the issue once we merge and the githash is stable.